### PR TITLE
Friendly name (#110)

### PR DESCRIPTION
**Pull request by zerodouglas**
Renames GetName to GetCardId. I added two new functions to the session interface so that the new resources to appropriately be called and added necessary changes to Swagger.

### CODE REVIEW

File: gui/api.go [hunk 1]
```diff
@@ -99,6 +99,8 @@ func Server(port string, certFile string, keyFile string, mock bool) {
 	r.HandleFunc("/cards/{sessionID}/init", session.init)
 	r.HandleFunc("/cards/{sessionID}/unlock", session.unlock)
 	r.HandleFunc("/cards/{sessionID}/pair", session.pair)
+	r.HandleFunc("/cards/{sessionID}/name", session.name)
```
Comment (npmaile) [inline gui/api.go:102]: just have the one endpoint of '/cards/{sessionID}/name'
use the verb to determine what to do. A GET should retrieve the name and a Post should set/overwrite the name

File: gui/api.go [hunk 2]
```diff
@@ -587,6 +589,58 @@ func (apiSession apiSession) pair(w http.ResponseWriter, r *http.Request) {
...
+		Name string
+	}{}
+	err = json.Unmarshal(body, &nameReq)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	err = sess.SetFriendlyName(nameReq.Name)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
```
Comment (npmaile) [inline gui/api.go:614]: just do a statusInternalServerError here though unprocessable entity may be more correct, but we would need to parse the error here to determine the precise issue before giving a more specific error response

File: gui/api.go [hunk 3]
```diff
@@ -587,6 +589,58 @@ func (apiSession apiSession) pair(w http.ResponseWriter, r *http.Request) {
...
+func (apiSession apiSession) name(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	sess, err := apiSession.sessionFromMuxVars(vars)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	name, err := sess.GetFriendlyName()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
```
Comment (npmaile) [inline gui/api.go:628]: same complaint as above

File: gui/api.go [hunk 4]
```diff
@@ -587,6 +589,58 @@ func (apiSession apiSession) pair(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	if name == "" {
+		http.Error(w, "No friendly name found", http.StatusNotFound)
```
Comment (npmaile) [inline gui/api.go:632]: after the change to cardID for the card's id, you can refer to the friendly name as "name"

**Review by npmaile — CHANGES_REQUESTED** [review]
You've done precisely what I wanted. Good work. You just need to change a couple of things around the http semantics and error messaging

File: gui/api.go [hunk 5]
```diff
@@ -587,6 +589,58 @@ func (apiSession apiSession) pair(w http.ResponseWriter, r *http.Request) {
 
 }
 
+func (apiSession apiSession) createName(w http.ResponseWriter, r *http.Request) {
```
Comment (maevyn11) [inline gui/api.go:592]: I'd prefer setName or setFriendlyName for this function. create implies a new name is being created, but this call could also change an existing friendly name setting.

File: gui/api.go [hunk 6]
```diff
@@ -587,6 +589,58 @@ func (apiSession apiSession) pair(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	err = sess.SetFriendlyName(nameReq.Name)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+		return
+	}
+}
+
+func (apiSession apiSession) name(w http.ResponseWriter, r *http.Request) {
```
Comment (maevyn11) [inline gui/api.go:619]: Would prefer getName or getFriendlyName for this to make it clear from outside what it's doing.

Comment (zerodouglas) [reply gui/api.go:619]: I suppose `getFriendlyName` is more clear about it's function.

File: gui/api.go (hunk 5 above)
Comment (zerodouglas) [reply gui/api.go:592]: I agree, I'll change to `setFriendlyName` to inline with `getFriendlyName`.

File: gui/api.go (hunk 4 above)
Comment (zerodouglas) [reply gui/api.go:632]: Sure, I'll get that changed to `sess.GetFriendName()` to `sess.GetName()` and `sess.SetName()`.

File: gui/swagger.yaml [hunk 7]
```diff
@@ -199,6 +199,46 @@ paths:
       type: string
       description: sessionID of connected card
 
+  /cards/{sessionID}/name:
+    get:
+      tags:
+      - "sessions"
+      responses:
+        "500":
```
Comment (maevyn11) [inline gui/swagger.yaml:207]: If you change the error code in the handler make sure this response code matches

**Comment by douglance** [comment]
Testing this out with swagger, it seems that after renaming a card, then calling `/listSessions` the `phonon-client` returns each session's original name, rather than the name it has been renamed to. The frontend will have to make another request for each session, which isn't a big deal; but we have two things with the same name. This seems a bit confusing. 

I would like to propose having `/listSessions` return `Id` as the initial identifier and `Name` as a string of it's friendly name or an empty string if it isn't set (it could also just be absent if not set).

So the change would be from 
```json
[{"Name":"043bf3fe28725a70","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
```
to:
```json
[{"Id":"043bf3fe28725a70","Name":"Friendly","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
```
or:
```json
[{"Id":"043bf3fe28725a70","Name":"","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
```

**Comment by zerodouglas** [comment]
> Testing this out with swagger, it seems that after renaming a card, then calling `/listSessions` the `phonon-client` returns each session's original name, rather than the name it has been renamed to. The frontend will have to make another request for each session, which isn't a big deal; but we have two things with the same name. This seems a bit confusing.
> 
> I would like to propose having `/listSessions` return `Id` as the initial identifier and `Name` as a string of it's friendly name or an empty string if it isn't set (it could also just be absent if not set).
> 
> So the change would be from
> 
> ```json
> [{"Name":"043bf3fe28725a70","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
> ```
> 
> to:
> 
> ```json
> [{"Id":"043bf3fe28725a70","Name":"Friendly","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
> ```
> 
> or:
> 
> ```json
> [{"Id":"043bf3fe28725a70","Name":"","Initialized":true,"TerminalPaired":true,"PinVerified":true}]
> ```
> 
> This would also make more sense with the naming of `GetCardId()` (if I understand the changes correctly (I do not write Go)).

I think this makes sense and not a hard feature to implement. @npmaile @maevyn11 should we cut the function to only act as a POST in this case?

**Comment by maevyn11** [comment]
Agree with @douglance 's suggestions as to renaming the fields passed to the frontend to disambiguate CardId and Name, passing both when available. 

Not sure what you mean by this? 

> should we cut the function to only act as a POST in this case?

Is the idea that the GET is unnecessary because the name will now be retrieved instead by /listSessions? 

I'm on board with that if so, better to have that functionality encapsulated in one place, but it does mean you'll need to ensure that the friendlyName will always have to be pulled into the session when it's available, so this call may need to be added to the session initialization sequence to ensure it's consistently cached in the client.

**Comment by zerodouglas** [comment]
> Agree with @douglance 's suggestions as to renaming the fields passed to the frontend to disambiguate CardId and Name, passing both when available.
> 
> Not sure what you mean by this?
> 
> > should we cut the function to only act as a POST in this case?
> 
> Is the idea that the GET is unnecessary because the name will now be retrieved instead by /listSessions?
> 
> I'm on board with that if so, better to have that functionality encapsulated in one place, but it does mean you'll need to ensure that the friendlyName will always have to be pulled into the session when it's available, so this call may need to be added to the session initialization sequence to ensure it's consistently cached in the client.

Yes, I had removed the GET request endpoint as it's now retrieved via /listSessions. I just made the changes to support your suggestion. The call will specifically return it from the session sequence.

**Comment by maevyn11** [comment]
All the code in the new changes looks good, but I think we should get the corresponding phonon-app changes committed to the submodule before merging, that way master will continue to function through the merge of the PR. https://github.com/GridPlus/phonon-app/pull/31 

Also conflicts with the other recent PRs need to be resolved before this one can merge.

**Review by maevyn11 — APPROVED** [review]
Tried it out and it works for me. Nicely done

**Comment by npmaile** [comment]
fix merge conflicts and ill test again and approve

**Comment by npmaile** [comment]
> fix merge conflicts and ill test again and approve

disregard. Force merging because I'll fix ci in a few minutes

**Pull request merged**